eZ/Publish/Core/Repository/ContentService.php

```diff
@@ -1294,7 +1294,8 @@ public function updateContent(APIVersionInfo $versionInfo, APIContentUpdateStruc
     protected function internalUpdateContent(
         APIVersionInfo $versionInfo,
         APIContentUpdateStruct $contentUpdateStruct,
-        ?array $fieldIdentifiersToValidate = null
+        ?array $fieldIdentifiersToValidate = null,
+        bool $doAddEmpty = false
```

Comment:
@alongosz I'll try to explain step by step what's going on behind the scenes.

After publishing version 2 having `eng-GB` and `ger-DE` translations we are at the following scenario:

 version | data_text | language_code
| ------------- | ------------- | -------------- |
| 1 |	one	| eng-GB |
| 1 |	two	| eng-GB |
| 2 |	one	| eng-GB |
| 2 |	two	| eng-GB |
| 2 |		| ger-DE |
| 2 |	one	| ger-DE |
| 3 |	one	| eng-GB |
| 3 |	two	| eng-GB |
| 3 |		| fre-FR |
| 3 |	one	| fre-FR |

Please note, that the 3rd version has only `eng-GB` and `fre-FR` translations because the 3rd version was created before we published the 2nd version.

Now, we publish the 3rd version with `eng-GB` and `fre-FR` languages.

 version | data_text | language_code
| ------------- | ------------- | -------------- |
| 1 |	one	| eng-GB |
| 1 |	two	| eng-GB |
| 2 |	one	| eng-GB |
| 2 |	two	| eng-GB |
| 2 |		| ger-DE |
| 2 |	one	| ger-DE |
| 3 |	one	| eng-GB |
| 3 |	two	| eng-GB |
| 3 |		| fre-FR |
| 3 |	one	| fre-FR |

Records are the same in both tables. Now, because the `ger-DE` translation is missing the `language_mask` in the `ezcontentobject` table is wrong because it contains only `eng-GB` and `fre-FR` translations, therefore, one - the `ger-DE` translation goes missing.

That's basically the issue from the ticket description.

Now, what we actually need to do to overcome this issue is somehow influence this `if`:
```php
if ($isRetained || $isCopied || ($isLanguageNew && $isEmpty && !$doAddEmpty) || $isProcessed) {
    continue;
}
```
and make the following conditional `false`: `($isLanguageNew && $isEmpty && !$doAddEmpty)` in case of empty fields because we would want those fields to be created. 

In the case of our scenario the `$isLanguageNew` variable is `true` and `isEmpty` is `true` as well, when publishing the 3rd version. In such a case the loop finishes at the `continue` preventing empty fields from creation.

Now to answer your questions:
> What makes me wonder though is how emptiness is related to the issue at hand. IBX-6833 descriptions states that in general, a translation is lost. In such case does this change solve the issue if it's applied to empty fields only? If we test this against more direct case mentioned in the ticket

The emptiness is the main issue here because the loop finishes at the `continue`. Currently in this PR **we are in fact applying this to empty fields only** because of `($isLanguageNew && $isEmpty && !$doAddEmpty)`.

> If we test this against more direct case mentioned in the ticket
Create a new content of the newly created Content Type with random values at two mentioned fields.
will that sill work somehow?

It will work properly if a translatable field is not empty. Only empty fields are causing issues.

> However, what are the side effects of always copying an empty field instead of just when copying translations?
I'm not sure if I understood you correctly but this won't make it work in case of always copying an empty field. The 3rd version does not have any information about the 2nd version, it's starting from 1st version therefore only the `copyTranslationsFromPublishedVersion` will allow us to get a necessary data from an earlier, published version.

It's kind of a complicated topic, if you wish we can sync regarding this.